core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -538,32 +551,52 @@ class SnappySession(_sc: SparkContext) extends SparkSession(_sc) {
     }
   }
 
+
+
   private[sql] def clearWriteLockOnTable(): Unit = {
+    logInfo(s"cleanWriteLockOnTable for app ${sqlContext.sparkContext.appName}",
+      new Throwable("ClearWrite"))
     contextObjects.remove(SnappySession.BULKWRITE_LOCK) match {
       case null =>
       case lock => releaseLock(lock)
     }
   }
 
   private[sql] def grabLock(table: String, schemaName: String,
-      connProperties: ConnectionProperties): Any = {
+      connProperties: ConnectionProperties): Option[Any] = {
     SnappyContext.getClusterMode(sparkContext) match {
       case _: ThinClientConnectorMode =>
+        while (!SnappySession.executorAssigned) {
+          // Don't expect this case usually unless lots of applications are submitted in parallel
+          try {
+            Thread.sleep(100)
```

Comment:
Should this also timeout? Otherwise while loop can be indefinite sometime?

Reply:
At spark level too we see that run job is blocked waiting. So, I thought indefinite wait is fine.
We can use wait/notify too.